src/wallet/walletdb.cpp

```diff
@@ -925,6 +928,7 @@ static DBErrors LoadDescriptorWalletRecords(CWallet* pwallet, DatabaseBatch& bat
                 return DBErrors::CORRUPT;
             }
             spk_man->AddKey(pubkey.GetID(), privkey);
+            desc_keys.insert(std::make_pair(std::make_pair(desc_id, pubkey.GetID()), privkey));
```

Comment:
Instead of adding all keys, which could be a high number, could only add the ones related to xpubs.
e.g. (quick and dirty, should also do the same for the encrypted ones)
```diff
diff --git a/src/wallet/walletdb.cpp b/src/wallet/walletdb.cpp
--- a/src/wallet/walletdb.cpp	(revision 7f44e83bb544f1e5c398b87c8eb7870eed3ed1ba)
+++ b/src/wallet/walletdb.cpp	(date 1690834804005)
@@ -903,10 +903,19 @@
         assert(spk_man);
         spk_man->SetCache(cache);
 
+        // Obtain xpub
+        std::set<CPubKey> desc_pubkeys;
+        std::set<CExtPubKey> desc_xpubs;
+        bool retain_xpub_privkey = true;
+        WITH_LOCK(spk_man->cs_desc_man, spk_man->GetWalletDescriptor().descriptor->GetPubKeys(desc_pubkeys, desc_xpubs));
+        if (desc_xpubs.size() != 1 || desc_pubkeys.size() != 0) {
+            retain_xpub_privkey = false;
+        }
+
         // Get unencrypted keys
         prefix = PrefixStream(DBKeys::WALLETDESCRIPTORKEY, id);
         LoadResult key_res = LoadRecords(pwallet, batch, DBKeys::WALLETDESCRIPTORKEY, prefix,
-            [&id, &spk_man, &desc_keys] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& strErr) {
+            [&id, &spk_man, &desc_keys, retain_xpub_privkey, &desc_xpubs] (CWallet* pwallet, DataStream& key, CDataStream& value, std::string& strErr) {
             uint256 desc_id;
             CPubKey pubkey;
             key >> desc_id;
@@ -936,7 +945,10 @@
                 return DBErrors::CORRUPT;
             }
             spk_man->AddKey(pubkey.GetID(), privkey);
-            desc_keys.insert(std::make_pair(std::make_pair(desc_id, pubkey.GetID()), privkey));
+
+            if (retain_xpub_privkey && desc_xpubs.begin()->pubkey == pubkey) {
+                desc_keys.insert(std::make_pair(std::make_pair(desc_id, pubkey.GetID()), privkey));
+            }
             return DBErrors::LOAD_OK;
         });
         result = std::max(result, key_res.m_result);
```

Reply:
I don't think this is a substantially useful optimization, and it still results in mixing upgrading code with the loading which I would rather not do.

Reply:
> I don't think this is a substantially useful optimization, and it still results in mixing upgrading code with the loading which I would rather not do.

As far as I remember, the last patch I shared doesn't mixes the upgrading code with the loading code.
It loads only the descriptor root key alone at the end of the spkm loading process so we can avoid loading all descriptors keys into a vector, just to obtain the private keys related to each descriptor xpub later in the upgrade process.

In other words, the difference is on a process who loads all keys in ram vs a process who loads only one key per descriptor.

That being said, my memory could be failing.. so np if you don't want to tackle it here. I could read what I did and revive it in a follow-up if it worth it.